File: backend/qa/alpha_js/lib/payment-terms.js
import { calcLegsForReinsurer, stripFacilityTag } from "lib/accounting.js";

export const MIN_PAYMENT_TERMS_DAYS = 15;

function num(value) {
  // Bugfix: Number("") is 0 (finite), not NaN, so an empty/absent value must be
  // rejected explicitly before the numeric conversion or every ?? fallback that
  // chains off this function (e.g. premium ?? ratio) silently never triggers.
  if (value === null || value === undefined) return null;
  const trimmed = String(value).trim();
  if (trimmed === "") return null;
  const n = Number(trimmed.replace(/,/g, ""));
  return Number.isFinite(n) ? n : null;
}

function money(value) {
  return Math.round((Number(value) + Number.EPSILON) * 100) / 100;
}

function validTerms(value) {
  const n = num(value);
  return n != null && Number.isInteger(n) && n >= MIN_PAYMENT_TERMS_DAYS ? n : null;
}

function validDate(value) {
  return /^\d{4}-\d{2}-\d{2}$/.test(String(value || ""));
}

export function addCalendarDays(value, days) {
  if (!validDate(value)) return "";
  const d = new Date(value + "T00:00:00Z");
  if (Number.isNaN(d.getTime())) return "";
  d.setUTCDate(d.getUTCDate() + Number(days || 0));
  return d.toISOString().slice(0, 10);
}

export function taipeiDate(now = new Date()) {
  return new Intl.DateTimeFormat("en-CA", {
    timeZone: "Asia/Taipei", year: "numeric", month: "2-digit", day: "2-digit"
  }).format(now);
}

export function daysBetweenDates(from, to) {
  if (!validDate(from) || !validDate(to)) return null;
  return Math.round((Date.parse(to + "T00:00:00Z") - Date.parse(from + "T00:00:00Z")) / 86400000);
}

function allocate(total, installments) {
  if (!installments.length) return [];
  const weights = installments.map((row) => Math.max(0, Number(row.weight) || 0));
  let sum = weights.reduce((a, b) => a + b, 0);
  if (sum <= 0) {
    sum = installments.length;
    weights.fill(1);
  }
  let used = 0;
  return installments.map((row, index) => {
    if (index === installments.length - 1) return money(total - used);
    const amount = money(total * weights[index] / sum);
    used = money(used + amount);
    return amount;
  });
}

export function paymentInstallments(caseData) {
  if (caseData?.installmentEnabled === true) {
    return (Array.isArray(caseData.performanceInstallments) ? caseData.performanceInstallments : []).map((row, index) => ({
      ...row,
      id: String(row?.id || "installment-" + (index + 1)),
      label: String(row?.performanceMonth || "Installment " + (index + 1)),
      baseDate: String(row?.paymentBaseDate || ""),
      weight: Math.max(0, num(row?.premium) ?? num(row?.ratio) ?? 0)
    }));
  }
  return [{
    id: "case-effective-date",
    label: "Case Effective Date",
    baseDate: String(caseData?.policyFrom || ""),
    weight: 1
  }];
}

function resolvedTerms(caseData, installment, reinsurer, reinsurerIndex) {
  const key = "r" + (reinsurerIndex + 1);
  const map = installment?.reinsurerPaymentTerms && typeof installment.reinsurerPaymentTerms === "object"
    ? installment.reinsurerPaymentTerms : {};
  return validTerms(map[key])
    ?? validTerms(reinsurer?.paymentTermsDays)
    ?? validTerms(installment?.paymentTermsDays)
    ?? validTerms(caseData?.paymentTermsDays);
}

function appliedByKey(caseData) {
  const result = {};
  const entries = Array.isArray(caseData?.paymentEntries) ? caseData.paymentEntries : [];
  entries.forEach((entry) => {
    const key = String(entry?.scheduleKey || "");
    const amount = num(entry?.amount);
    if (!key || amount == null) return;
    const signed = entry?.entryType === "reversal" ? -Math.abs(amount) : Math.abs(amount);
    result[key] = money((result[key] || 0) + signed);
  });
  return result;
}

function statusFor(dueDate, outstanding, paid, today, now) {
  if (!dueDate) return "pending_review";
  if (outstanding <= 0.004) return "settled";
  const delta = daysBetweenDates(dueDate, today);
  if (delta > 0) return "overdue";
  if (delta === 0) {
    const time = new Intl.DateTimeFormat("en-GB", {
      timeZone: "Asia/Taipei", hour: "2-digit", minute: "2-digit", hour12: false
    }).format(now);
    return time > "12:00" ? "overdue" : "due_today";
  }
  return paid > 0.004 ? "partially_paid" : "upcoming";
}

export function buildPaymentSchedule(caseData, now = new Date()) {
  const reinsurers = Array.isArray(caseData?.reinsurers) ? caseData.reinsurers : [];
  const installments = paymentInstallments(caseData);
  const applied = appliedByKey(caseData);
  const today = taipeiDate(now);
  const issues = [];
  const weights = installments.map((row) => ({ ...row, weight: row.weight }));
  const legs = reinsurers.map((reinsurer) => calcLegsForReinsurer(caseData, reinsurer));
  const cedantAmounts = allocate(legs.reduce((sum, row) => sum + row.leg1, 0), weights);
  const reinsurerAmounts = legs.map((row) => allocate(row.leg2, weights));
  const items = [];

  if (validTerms(caseData?.paymentTermsDays) == null) issues.push("Case Payment Terms must be at least 15 calendar days.");

  installments.forEach((installment, installmentIndex) => {
    const baseDate = validDate(installment.baseDate) ? installment.baseDate : "";
    if (!baseDate) issues.push(installment.label + ": payment base date is required.");

    const resolvedTermsPerReinsurer = reinsurers.map((reinsurer, index) => resolvedTerms(caseData, installment, reinsurer, index));
    const allTerms = resolvedTermsPerReinsurer.filter((value) => value != null);
    // Low/observational fix: the cedant due date below takes the min() of only the
    // VALID reinsurer terms -- any invalid one is silently excluded from that
    // calculation. A per-reinsurer "invalid Payment Terms" issue is still raised
    // separately further down, but it doesn't say it also affected this cedant
    // calculation, so make that link explicit here.
    if (allTerms.length < resolvedTermsPerReinsurer.length) {
      issues.push(installment.label + ": Cedant due date calculation excludes Reinsurer(s) with invalid Payment Terms.");
    }
    const cedantTerms = allTerms.length ? Math.min(...allTerms) : validTerms(installment?.paymentTermsDays) ?? validTerms(caseData?.paymentTermsDays);
    const cedantKey = installment.id + ":cedant";
    const cedantPaid = Math.max(0, money(applied[cedantKey] || 0));
    const cedantAmount = Math.max(0, money(cedantAmounts[installmentIndex] || 0));
    const cedantOutstanding = Math.max(0, money(cedantAmount - cedantPaid));
    const cedantDue = baseDate && cedantTerms != null ? addCalendarDays(baseDate, cedantTerms - 15) : "";
    items.push({
      scheduleKey: cedantKey, installmentId: installment.id, installmentLabel: installment.label,
      partyType: "cedant", partyKey: "cedant", partyName: String(caseData?.reinsuredName || caseData?.cedantName || "Cedant"),
      currency: String(caseData?.currency || ""), baseDate, termsDays: cedantTerms,
      dueDate: cedantDue, dueTime: "12:00", amount: cedantAmount, paid: Math.min(cedantAmount, cedantPaid),
      outstanding: cedantOutstanding, status: statusFor(cedantDue, cedantOutstanding, cedantPaid, today, now),
      partial: cedantPaid > 0.004 && cedantOutstanding > 0.004
    });

    reinsurers.forEach((reinsurer, reinsurerIndex) => {
      const terms = resolvedTerms(caseData, installment, reinsurer, reinsurerIndex);
      if (terms == null) issues.push(installment.label + ": invalid Payment Terms for Reinsurer " + (reinsurerIndex + 1) + ".");
      const key = installment.id + ":reinsurer:r" + (reinsurerIndex + 1);
      const paid = Math.max(0, money(applied[key] || 0));
      const amount = Math.max(0, money(reinsurerAmounts[reinsurerIndex]?.[installmentIndex] || 0));
      const outstanding = Math.max(0, money(amount - paid));
      const dueDate = baseDate && terms != null ? addCalendarDays(baseDate, terms - 10) : "";
      items.push({
        scheduleKey: key, installmentId: installment.id, installmentLabel: installment.label,
        partyType: "reinsurer", partyKey: "r" + (reinsurerIndex + 1),
        partyName: String(reinsurer?.foreignBroker || stripFacilityTag(reinsurer?.name) || "Reinsurer " + (reinsurerIndex + 1)),
        currency: String(caseData?.currency || ""), baseDate, termsDays: terms,
        dueDate, dueTime: "12:00", amount, paid: Math.min(amount, paid), outstanding,
        status: statusFor(dueDate, outstanding, paid, today, now),
        partial: paid > 0.004 && outstanding > 0.004
      });
    });
  });

  return {
    items,
    issues: [...new Set(issues)],
    reviewRequired: issues.some((issue) => issue.includes("payment base date")),
    totals: {
      cedantOutstanding: money(items.filter((row) => row.partyType === "cedant").reduce((sum, row) => sum + row.outstanding, 0)),
      reinsurerOutstanding: money(items.filter((row) => row.partyType === "reinsurer").reduce((sum, row) => sum + row.outstanding, 0))
    }
  };
}

// Maps a single row of the OLD premium ledger (payload.transactions[], produced
// by lib/accounting.js buildPremiumTransactions at Production Report close) to a
// settlement status derived from the NEW payment schedule (paymentEntries[]).
// The old ledger's own `settlement` field is permanently stale for premium rows
// (nothing has written to it since Payment Terms shipped), so this recomputes a
// status on the fly instead. Claim-sourced rows are untouched — a different,
// still-working code path keeps their `settlement` field accurate.
//
// Leg 1 (cedant receivable, per-reinsurer allocation in the old ledger) maps to
// the case's single pooled cedant obligation — the cedant pays once for the
// whole case, not per reinsurer, so every Leg 1 row (including split a/b) shares
// one true status.
// Leg 2 (reinsurer payable) maps cleanly to that specific reinsurer's schedule
// rows, summed across installments — the underlying amount formula is identical,
// just no longer collapsed into one lump sum.
// Leg 3 (brokerage spread) has no equivalent in the payment schedule at all — it
// was never turned into a trackable, payable line — so it is reported as
// "not_tracked" rather than a possibly-wrong open/settled guess.
// Performance Split a/b rows for the same reinsurer/leg share one status, since
// the payment schedule has no concept of split parties.
export function deriveLedgerSettlement(caseData, transaction) {
  if (transaction?.source === 'claim') {
    return transaction?.settlement === 'settled' ? 'settled' : 'open';
  }
  const legType = String(transaction?.legType || '');
  const schedule = buildPaymentSchedule(caseData);
  const summarize = (items) => {
    if (!items.length) return 'not_tracked';
    const outstanding = items.reduce((sum, item) => sum + item.outstanding, 0);
    const paid = items.reduce((sum, item) => sum + item.paid, 0);
    if (outstanding <= 0.004) return 'settled';
    if (paid > 0.004) return 'partial';
    return 'open';
  };
  if (legType.startsWith('Leg 1')) {
    return summarize(schedule.items.filter((item) => item.partyType === 'cedant'));
  }
  if (legType.startsWith('Leg 2')) {
    const reinsurerIdx = Number.isInteger(transaction?.reinsurerIdx) ? transaction.reinsurerIdx : -1;
    if (reinsurerIdx < 0) return 'not_tracked';
    const key = 'r' + (reinsurerIdx + 1);
    return summarize(schedule.items.filter((item) => item.partyType === 'reinsurer' && item.partyKey === key));
  }
  return 'not_tracked';
}

export function reminderKind(item, today = taipeiDate()) {
  if (!item?.dueDate || item.outstanding <= 0.004) return "";
  const days = daysBetweenDates(today, item.dueDate);
  if (days === 7) return "seven_days_before";
  if (days === 0) return "due_today";
  const overdueDays = daysBetweenDates(item.dueDate, today);
  if (overdueDays > 0 && overdueDays % 7 === 0) return "weekly_overdue";
  return "";
}

File: backend/cases/calc/payment_terms.py
"""
付款期限、付款排程與帳款結清狀態 - 逐行移植自 Hatchable Alpha 的 lib/payment-terms.js（v53）。

與 accounting.py 一樣，刻意保留 JavaScript 的語意，並由差異測試對 Alpha 的 JavaScript 逐位比對。
Alpha 的函數裡有兩處會讀「現在時間」（buildPaymentSchedule、deriveLedgerSettlement 的預設值），
這裡把 now 做成可傳入的參數，測試才能固定時間；正式使用時不傳就是現在。
"""
import math
import re
from datetime import datetime, timezone
from zoneinfo import ZoneInfo

from .accounting import calc_legs_for_reinsurer, strip_facility_tag
from .jsnum import (
    UNDEFINED, at, civil_from_days, get, is_array, js_spread, is_nullish, iso_date_prefix, js_is_finite, js_is_integer, js_or,
    js_max, js_min, js_round, js_string_to_number, js_to_number, js_to_string, js_truthy, js_trim, money,
    parse_iso_date_days, MAX_DAYS,
)

MIN_PAYMENT_TERMS_DAYS = 15
TAIPEI = ZoneInfo("Asia/Taipei")
_DATE_SHAPE = re.compile(r"[0-9]{4}-[0-9]{2}-[0-9]{2}\Z")


def num(value):
    """空值或空字串回傳 None（不是 0）；其餘依 Number() 解析，非有限數字回傳 None。"""
    if is_nullish(value):
        return None
    trimmed = js_trim(js_to_string(value))
    if trimmed == "":
        return None
    n = js_string_to_number(trimmed.replace(",", ""))
    return n if js_is_finite(n) else None


def valid_terms(value):
    n = num(value)
    return n if n is not None and js_is_integer(n) and n >= MIN_PAYMENT_TERMS_DAYS else None


def valid_date(value):
    return _DATE_SHAPE.match(js_to_string(js_or(value, ""))) is not None


def add_calendar_days(value, days):
    """回傳 YYYY-MM-DD 字串；日期格式不對回傳 ""。與 Alpha 相同，天數不是有限數字時會丟出例外（Date.toISOString 的 RangeError）。"""
    if not valid_date(value):
        return ""
    base = parse_iso_date_days(value)
    if base is None:
        return ""
    n = js_to_number(js_or(days, 0))
    day_of_month = civil_from_days(base)[2]
    target = day_of_month + n              # setUTCDate 收到的是「當月第幾日 + 天數」，再取整數部分
    if not js_is_finite(target):
        raise ValueError("Invalid time value")
    total = base + (int(math.trunc(target)) - day_of_month)
    if abs(total) > MAX_DAYS:
        raise ValueError("Invalid time value")
    return iso_date_prefix(total)


def _as_utc(now):
    if now is None:
        return datetime.now(timezone.utc)
    if isinstance(now, str):
        return datetime.fromisoformat(now.replace("Z", "+00:00"))
    return now


def taipei_date(now=None):
    return _as_utc(now).astimezone(TAIPEI).strftime("%Y-%m-%d")


def days_between_dates(date_from, date_to):
    """相差天數；任一邊不是有效日期回傳 None（Alpha 這裡回傳 null）；日期滾動後無效則回傳 NaN。"""
    if not valid_date(date_from) or not valid_date(date_to):
        return None
    a, b = parse_iso_date_days(date_from), parse_iso_date_days(date_to)
    if a is None or b is None:
        return math.nan
    return js_round((b - a) * 86400000 / 86400000)


def allocate(total, installments):
    if not installments:
        return []
    weights = []
    for row in installments:
        w = js_to_number(row["weight"])
        weights.append(js_max(0.0, 0.0 if (w != w or w == 0) else w))   # Math.max(0, Number(w) || 0)
    total_weight = 0.0
    for w in weights:
        total_weight += w
    if total_weight <= 0:
        total_weight = float(len(installments))
        weights = [1.0] * len(installments)
    used = 0.0
    result = []
    last = len(installments) - 1
    for index, _row in enumerate(installments):
        if index == last:
            result.append(money(total - used))
        else:
            amount = money(total * weights[index] / total_weight)
            used = money(used + amount)
            result.append(amount)
    return result


def payment_installments(case_data):
    if get(case_data, "installmentEnabled") is True:
        rows = get(case_data, "performanceInstallments")
        rows = rows if is_array(rows) else []
        result = []
        for index, row in enumerate(rows):
            entry = js_spread(row)
            weight = num(get(row, "premium"))
            if weight is None:
                weight = num(get(row, "ratio"))
            if weight is None:
                weight = 0
            entry.update({
                "id": js_to_string(js_or(get(row, "id"), "installment-" + str(index + 1))),
                "label": js_to_string(js_or(get(row, "performanceMonth"), "Installment " + str(index + 1))),
                "baseDate": js_to_string(js_or(get(row, "paymentBaseDate"), "")),
                "weight": js_max(0.0, float(weight)),
            })
            result.append(entry)
        return result
    return [{
        "id": "case-effective-date",
        "label": "Case Effective Date",
        "baseDate": js_to_string(js_or(get(case_data, "policyFrom"), "")),
        "weight": 1,
    }]


def _resolved_terms(case_data, installment, reinsurer, reinsurer_index):
    key = "r" + str(reinsurer_index + 1)
    raw = get(installment, "reinsurerPaymentTerms")
    mapping = raw if (js_truthy_object(raw)) else {}
    for candidate in (
        get(mapping, key) if isinstance(mapping, dict) else UNDEFINED,
        get(reinsurer, "paymentTermsDays"),
        get(installment, "paymentTermsDays"),
        get(case_data, "paymentTermsDays"),
    ):
        terms = valid_terms(candidate)
        if terms is not None:
            return terms
    return None


def _or_zero(value):
    """x || 0"""
    return value if js_truthy(value) else 0


def js_truthy_object(value):
    """value && typeof value === 'object'（陣列與物件都算；null 不算）"""
    return isinstance(value, (dict, list))


def _applied_by_key(case_data):
    result = {}
    entries = get(case_data, "paymentEntries")
    entries = entries if is_array(entries) else []
    for entry in entries:
        key = js_to_string(js_or(get(entry, "scheduleKey"), ""))
        amount = num(get(entry, "amount"))
        if not key or amount is None:
            continue
        signed = -abs(amount) if get(entry, "entryType") == "reversal" else abs(amount)
        previous = result.get(key)
        result[key] = money((previous if js_truthy(previous) else 0) + signed)
    return result


def _status_for(due_date, outstanding, paid, today, now):
    if not due_date:
        return "pending_review"
    if outstanding <= 0.004:
        return "settled"
    delta = days_between_dates(due_date, today)
    if delta is not None and delta > 0:
        return "overdue"
    if delta == 0:
        time = _as_utc(now).astimezone(TAIPEI).strftime("%H:%M")
        return "overdue" if time > "12:00" else "due_today"
    return "partially_paid" if paid > 0.004 else "upcoming"


def build_payment_schedule(case_data, now=None):
    now = _as_utc(now)
    reinsurers = get(case_data, "reinsurers")
    reinsurers = reinsurers if is_array(reinsurers) else []
    installments = payment_installments(case_data)
    applied = _applied_by_key(case_data)
    today = taipei_date(now)
    issues = []
    weights = [{**row, "weight": row["weight"]} for row in installments]
    legs = [calc_legs_for_reinsurer(case_data, r) for r in reinsurers]
    leg1_total = 0.0
    for row in legs:
        leg1_total += row["leg1"]
    cedant_amounts = allocate(leg1_total, weights)
    reinsurer_amounts = [allocate(row["leg2"], weights) for row in legs]
    items = []

    if valid_terms(get(case_data, "paymentTermsDays")) is None:
        issues.append("Case Payment Terms must be at least 15 calendar days.")

    for installment_index, installment in enumerate(installments):
        base_date = installment["baseDate"] if valid_date(installment["baseDate"]) else ""
        if not base_date:
            issues.append(installment["label"] + ": payment base date is required.")

        resolved = [_resolved_terms(case_data, installment, r, i) for i, r in enumerate(reinsurers)]
        all_terms = [t for t in resolved if t is not None]
        if len(all_terms) < len(resolved):
            issues.append(installment["label"] + ": Cedant due date calculation excludes Reinsurer(s) with invalid Payment Terms.")
        if all_terms:
            cedant_terms = js_min(*all_terms)
        else:
            cedant_terms = valid_terms(get(installment, "paymentTermsDays"))
            if cedant_terms is None:
                cedant_terms = valid_terms(get(case_data, "paymentTermsDays"))
        cedant_key = installment["id"] + ":cedant"
        cedant_paid = js_max(0.0, money(_or_zero(applied.get(cedant_key))))
        cedant_amount = js_max(0.0, money(_or_zero(cedant_amounts[installment_index] if installment_index < len(cedant_amounts) else None)))
        cedant_outstanding = js_max(0.0, money(cedant_amount - cedant_paid))
        cedant_due = add_calendar_days(base_date, cedant_terms - 15) if base_date and cedant_terms is not None else ""
        items.append({
            "scheduleKey": cedant_key, "installmentId": installment["id"], "installmentLabel": installment["label"],
            "partyType": "cedant", "partyKey": "cedant",
            "partyName": js_to_string(js_or(get(case_data, "reinsuredName"), get(case_data, "cedantName"), "Cedant")),
            "currency": js_to_string(js_or(get(case_data, "currency"), "")), "baseDate": base_date, "termsDays": cedant_terms,
            "dueDate": cedant_due, "dueTime": "12:00", "amount": cedant_amount, "paid": js_min(cedant_amount, cedant_paid),
            "outstanding": cedant_outstanding, "status": _status_for(cedant_due, cedant_outstanding, cedant_paid, today, now),
            "partial": cedant_paid > 0.004 and cedant_outstanding > 0.004,
        })

        for reinsurer_index, reinsurer in enumerate(reinsurers):
            terms = _resolved_terms(case_data, installment, reinsurer, reinsurer_index)
            if terms is None:
                issues.append(installment["label"] + ": invalid Payment Terms for Reinsurer " + str(reinsurer_index + 1) + ".")
            key = installment["id"] + ":reinsurer:r" + str(reinsurer_index + 1)
            paid = js_max(0.0, money(_or_zero(applied.get(key))))
            row_amounts = reinsurer_amounts[reinsurer_index] if reinsurer_index < len(reinsurer_amounts) else []
            amount = js_max(0.0, money(_or_zero(row_amounts[installment_index] if installment_index < len(row_amounts) else None)))
            outstanding = js_max(0.0, money(amount - paid))
            due_date = add_calendar_days(base_date, terms - 10) if base_date and terms is not None else ""
            items.append({
                "scheduleKey": key, "installmentId": installment["id"], "installmentLabel": installment["label"],
                "partyType": "reinsurer", "partyKey": "r" + str(reinsurer_index + 1),
                "partyName": js_to_string(js_or(get(reinsurer, "foreignBroker"), strip_facility_tag(get(reinsurer, "name")),
                                                "Reinsurer " + str(reinsurer_index + 1))),
                "currency": js_to_string(js_or(get(case_data, "currency"), "")), "baseDate": base_date, "termsDays": terms,
                "dueDate": due_date, "dueTime": "12:00", "amount": amount, "paid": js_min(amount, paid), "outstanding": outstanding,
                "status": _status_for(due_date, outstanding, paid, today, now),
                "partial": paid > 0.004 and outstanding > 0.004,
            })

    cedant_out = 0.0
    reinsurer_out = 0.0
    for row in items:
        if row["partyType"] == "cedant":
            cedant_out += row["outstanding"]
    for row in items:
        if row["partyType"] == "reinsurer":
            reinsurer_out += row["outstanding"]
    return {
        "items": items,
        "issues": list(dict.fromkeys(issues)),
        "reviewRequired": any("payment base date" in issue for issue in issues),
        "totals": {"cedantOutstanding": money(cedant_out), "reinsurerOutstanding": money(reinsurer_out)},
    }


def derive_ledger_settlement(case_data, transaction, now=None):
    """
    舊帳本（payload.transactions[]）的一筆交易，對應到新付款排程（paymentEntries[]）的結清狀態。
    Leg 1 對應 Cedant；Leg 2 對應該再保人；Leg 3（佣金差）沒有對應的排程，回報 not_tracked。
    """
    if get(transaction, "source") == "claim":
        return "settled" if get(transaction, "settlement") == "settled" else "open"
    leg_type = js_to_string(js_or(get(transaction, "legType"), ""))
    schedule = build_payment_schedule(case_data, now)

    def summarize(items):
        if not items:
            return "not_tracked"
        outstanding = 0.0
        paid = 0.0
        for item in items:
            outstanding += item["outstanding"]
        for item in items:
            paid += item["paid"]
        if outstanding <= 0.004:
            return "settled"
        if paid > 0.004:
            return "partial"
        return "open"

    if leg_type.startswith("Leg 1"):
        return summarize([i for i in schedule["items"] if i["partyType"] == "cedant"])
    if leg_type.startswith("Leg 2"):
        idx = get(transaction, "reinsurerIdx")
        reinsurer_idx = int(idx) if js_is_integer(idx) else -1
        if reinsurer_idx < 0:
            return "not_tracked"
        key = "r" + str(reinsurer_idx + 1)
        return summarize([i for i in schedule["items"] if i["partyType"] == "reinsurer" and i["partyKey"] == key])
    return "not_tracked"


def reminder_kind(item, today=None):
    due = get(item, "dueDate")
    if not js_truthy(due):
        return ""
    if js_to_number(get(item, "outstanding")) <= 0.004:   # undefined <= 0.004 是 false（繼續往下）；null 當 0
        return ""
    today = today if today is not None else taipei_date()
    days = days_between_dates(today, due)
    if days == 7:
        return "seven_days_before"
    if days == 0:
        return "due_today"
    overdue = days_between_dates(due, today)
    if overdue is not None and overdue > 0 and overdue % 7 == 0:
        return "weekly_overdue"
    return ""
